File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        ans = 0
        def dfs(node, last):
            nonlocal ans
            if node.val>=last:
                ans += 1
            if not node.left and not node.right:
                return
            last = max(node.val, last)
            if node.left:
                dfs(node.left, last)
            if node.right:
                dfs(node.right, last)
        
        dfs(root, float('-inf'))

        return ans
            

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int ans = 0;
        DFS(root, ans);
        return ans;

    }
private:
    int DFS(TreeNode* node, int &ans){
        if(node == nullptr) return 0;
        int lh = DFS(node->left, ans);
        int rh = DFS(node->right, ans); 

        ans = max(ans, lh+rh);
        return max(lh, rh)+1;
    }
};
